### ISSUE TITLE
Different outputs on isomorphic graphs?

### ISSUE THREAD

**Issue by anthonimes** [LyteFM/modular-decomposition#2]
When considering the following graph (the displayed multiple edges are only one edge):
![u_44](https://user-images.githubusercontent.com/2004010/54825298-853e0c00-4cad-11e9-9457-b7ef69c1c222.png)

the MD tree obtained contains a large Series node, which does not seem to exist in the graph
![MD_u_44](https://user-images.githubusercontent.com/2004010/54825295-84a57580-4cad-11e9-8412-93d8424cca13.png)

The graph in the string format: 

`std::string u_44 = "([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 14, 15, 16, 17, 18, 19], [{0, 8}, {0, 19}, {1, 2}, {1, 3}, {1, 4}, {1, 7}, {1, 9}, {1, 13}, {1, 14}, {2, 4}, {2, 9}, {2, 13}, {2, 14}, {2, 17}, {3, 8}, {4, 13}, {5, 8}, {6, 8}, {7, 8}, {7, 10}, {7, 11}, {8, 15}, {8, 16}, {8, 17}, {8, 18}, {13, 14}])";`

An interesting point is that, if we consider the ` u_27.ed` (which is isomorphic to u_44), the MD obtained seems more relevant: 
![u_27](https://user-images.githubusercontent.com/2004010/54825296-84a57580-4cad-11e9-8814-e39b356eb618.png)
![MD_u_27](https://user-images.githubusercontent.com/2004010/54825294-84a57580-4cad-11e9-998b-2a334c77dd7a.png)

Again, the graph in string format: 

`std::string u_27 = "([0, 1, 2, 3, 4, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19], [{0, 1}, {0, 2}, {0, 3}, {0, 4}, {0, 6}, {0, 10}, {0, 18}, {1, 2}, {1, 3}, {1, 6}, {1, 10}, {1, 17}, {2, 3}, {2, 6}, {4, 19}, {7, 19}, {8, 19}, {9, 19}, {11, 19}, {12, 19}, {13, 18}, {14, 16}, {15, 18}, {16, 19}, {17, 19}, {18, 19}])";` 

Is it possible that the order in which the vertices are considered impact the output?

**Comment by fynnlyte** [comment]
Yeah, the modular decomposition should definitely be the same for isomorphic graphs, this is definitely a bug. And since it even classified a prime graph as non-prime this also affects the code that is based on it. I hope I'll find time to look into in this weekend.
@AlphaSquad have you ever noticed something like this?

**Comment by anthonimes** [comment]
Thanks for the answer! I started reading the paper from Capelle et al. to see if I can see something, but I have quite a long road to take before finding anything. :) 

If I can help you in any -other- way, please let me know!

**Comment by fynnlyte** [comment]
You could try the following to identify the bug:
- Check if the error still persists when using the original code which included the labelling of spider modules. I might have done another mistake when changing the code. You can find it [here](https://github.com/LyteFM/modular-decomposition/tree/original-spider)
- Print the Dyck Word (the bracket expression), you just need to uncomment line 624, and check if the structure is matching - they do not need to be exactly the same, but should have the same sub-bracket structures.
- Print the number of edges and vertices in the imported graph - or even better, check if they still are isomorphic after the import -  to exclude that the bug is happening already there.

**Comment by AlphaSquad** [comment]
Hey,
nice to see that this is being used - I didn't know that! I ran your graph through my original code and got the following MD tree:
![graph dot](https://user-images.githubusercontent.com/11226638/55222437-33950480-520c-11e9-8b24-030b81f5bd7d.png)
>The vertex names seem...off, but the tree looks isomorphic to your second one, so I'm guessing I just did some error with I/O, because my version does/did not support the graph as string so I had to type it in manually.

E: I just noticed that it should be correct with the vertex naming of the first graph, I only compared to the second one, which obviously has different vertex names

**Comment by AlphaSquad** [comment]
Also, it worked for me with the given code. I noticed though that it is a problem if there is a space between the two vertices of an edge in the string, i.e. 
`./mod_dec "([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 14, 15, 16, 17, 18, 19], [{0, 8}, {0, 19}, {1, 2}, {1, 3}, {1, 4}, {1, 7}, {1, 9}, {1, 13}, {1, 14}, {2, 4}, {2, 9}, {2, 13}, {2, 14}, {2, 17}, {3, 8}, {4, 13}, {5, 8}, {6, 8}, {7, 8}, {7, 10}, {7, 11}, {8, 15}, {8, 16}, {8, 17}, {8, 18}, {13, 14}])"`
would lead to a lot of errors in parsing (and resulting in one parallel node because all edges were missing), while
`./mod_dec "([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 14, 15, 16, 17, 18, 19], [{0,8}, {0,19}, {1,2}, {1,3}, {1,4}, {1,7}, {1,9}, {1,13}, {1,14}, {2,4}, {2,9}, {2,13}, {2,14}, {2,17}, {3,8}, {4,13}, {5,8}, {6,8}, {7,8}, {7,10}, {7,11}, {8,15}, {8,16}, {8,17}, {8,18}, {13,14}])"`
worked perfectly:
![graph dot](https://user-images.githubusercontent.com/11226638/55223265-182af900-520e-11e9-9a44-c2df00dba04d.png)
Could you guys also try this again? Can you reproduce?

**Comment by AlphaSquad** [comment]
Aha! I have been playing a bit and it has to do with your double-edges.
Running the code with the duplicated edges results in your graph:
`./mod_dec "([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 14, 15, 16, 17, 18, 19], [{0,8}, {0,19}, {1,2}, {1,3}, {1,4}, {1,7}, {1,9}, {1,13}, {1,14}, {2,4}, {2,9}, {2,13}, {2,14}, {2,17}, {3,8}, {4,13}, {5,8}, {6,8}, {7,8}, {7,10}, {7,11}, {8,15}, {8,16}, {8,17}, {8,18}, {13,14}, {0,8}, {0,19}, {1,2}, {1,3}, {1,4}, {1,7}, {1,9}, {1,13}, {1,14}, {2,4}, {2,9}, {2,13}, {2,14}, {2,17}, {3,8}, {4,13}, {5,8}, {6,8}, {7,8}, {7,10}, {7,11}, {8,15}, {8,16}, {8,17}, {8,18}, {13,14}])"`
![graph2 dot](https://user-images.githubusercontent.com/11226638/55224988-e87df000-5211-11e9-8ea1-adb2478c0b7a.png)
Interestingly enough, using my original code, the MD tree gains another prime vertex which has no children though. Peculiar, but I think the algorithm is just not suitable for multi-graphs

**Comment by anthonimes** [comment]
Oh I think you're right! I dumbly assumed that multi-edges would not be an issue, but removing them indeed seems to do the trick! 

I have further tests to make I'll let you know if something odd happens again. 

Thanks for both your code and assistance, really helpful!
